File: src/board/board.ts
import { STROKE } from '../config';
import type { FrameMapper } from '../tracking/frameMapper';
import type { BoardAction, BoardPoint, StrokeAction, Tool } from '../types';

const TAU = Math.PI * 2;

/**
 * Traces a stroke as a chain of quadratic curves whose control points are the
 * samples and whose endpoints are the midpoints between them. Two things fall
 * out of that: the curve is C1 continuous, so the line reads as one gesture
 * rather than a polyline, and widely spaced samples during a fast movement are
 * bridged by a curve instead of leaving a gap.
 */
function tracePath(ctx: CanvasRenderingContext2D, stroke: StrokeAction, mapper: FrameMapper): void {
  const pts = stroke.points;
  const n = pts.length;
  const width = Math.max(stroke.width * mapper.pixelsPerUnit, 0.6);

  if (n === 1) {
    const p = mapper.boardToStage(pts[0].x, pts[0].y);
    ctx.beginPath();
    ctx.arc(p.x, p.y, width / 2, 0, TAU);
    ctx.fill();
    return;
  }

  ctx.lineWidth = width;
  ctx.beginPath();

  const first = mapper.boardToStage(pts[0].x, pts[0].y);
  ctx.moveTo(first.x, first.y);

  if (n === 2) {
    const last = mapper.boardToStage(pts[1].x, pts[1].y);
    ctx.lineTo(last.x, last.y);
  } else {
    let current = mapper.boardToStage(pts[1].x, pts[1].y);
    for (let i = 1; i < n - 1; i += 1) {
      const next = mapper.boardToStage(pts[i + 1].x, pts[i + 1].y);
      ctx.quadraticCurveTo(current.x, current.y, (current.x + next.x) / 2, (current.y + next.y) / 2);
      current = next;
    }
    ctx.lineTo(current.x, current.y);
  }

  ctx.stroke();
}

function paintStroke(
  ctx: CanvasRenderingContext2D,
  stroke: StrokeAction,
  mapper: FrameMapper,
): void {
  ctx.save();
  ctx.lineCap = 'round';
  ctx.lineJoin = 'round';

  if (stroke.tool === 'eraser') {
    // Erasing punches a hole in the ink layer rather than painting background
    // over it, which is what keeps the whiteboard transparent over the video.
    ctx.globalCompositeOperation = 'destination-out';
    ctx.strokeStyle = '#000';
    ctx.fillStyle = '#000';
  } else {
    ctx.globalCompositeOperation = 'source-over';
    ctx.strokeStyle = stroke.color;
    ctx.fillStyle = stroke.color;
  }

  tracePath(ctx, stroke, mapper);
  ctx.restore();
}

export function paintActions(
  ctx: CanvasRenderingContext2D,
  actions: readonly BoardAction[],
  mapper: FrameMapper,
  width: number,
  height: number,
): void {
  ctx.clearRect(0, 0, width, height);
  for (const action of actions) {
    if (action.kind === 'clear') ctx.clearRect(0, 0, width, height);
    else paintStroke(ctx, action, mapper);
  }
}

export interface BoardStats {
  strokes: number;
  canUndo: boolean;
  canRedo: boolean;
}

/**
 * Holds the drawing as a list of vector actions and keeps a rasterised cache of
 * the finished ones. Finished strokes are painted into the cache exactly once;
 * only in-progress strokes are re-traced each frame. Undo, redo, clear and
 * window resizes replay the action list, which is also what makes undo work
 * across an erase or a clear.
 */
export class Board {
  private actions: BoardAction[] = [];
  private redoStack: BoardAction[] = [];
  private readonly active = new Map<number, StrokeAction>();

  private readonly cache: HTMLCanvasElement;
  private readonly cacheCtx: CanvasRenderingContext2D;
  private cacheStale = true;
  /** The mapper the cache was rasterised with; null until the first render. */
  private cacheMapper: FrameMapper | null = null;
  /** Cache dimensions in CSS pixels — its context carries the DPR transform. */
  private cacheWidth = 0;
  private cacheHeight = 0;

  /** Bumped whenever the visible result would differ — the app uses it to skip redraws. */
  version = 0;

  onChange: ((stats: BoardStats) => void) | null = null;

  constructor() {
    this.cache = document.createElement('canvas');
    const ctx = this.cache.getContext('2d');
    if (!ctx) throw new Error('Canvas 2D is unavailable in this browser.');
    this.cacheCtx = ctx;
  }

  /**
   * `strokes` counts what is currently *visible*, not everything ever drawn —
   * a clear resets it to zero. Clear and Save key off this, and both should be
   * dead on an empty board even though the history behind it is still intact.
   */
  get stats(): BoardStats {
    let visible = 0;
    for (const action of this.actions) {
      if (action.kind === 'clear') visible = 0;
      else visible += 1;
    }
    return {
      strokes: visible,
      canUndo: this.actions.length > 0,
      canRedo: this.redoStack.length > 0,
    };
  }

  get activeStrokes(): IterableIterator<StrokeAction> {
    return this.active.values();
  }

  get hasActiveStrokes(): boolean {
    return this.active.size > 0;
  }

  hasActiveStroke(handId: number): boolean {
    return this.active.has(handId);
  }

  getActions(): readonly BoardAction[] {
    return this.actions;
  }

  /** @param dpr device pixel ratio — the cache renders at full device resolution. */
  resize(cssWidth: number, cssHeight: number, dpr: number): void {
    const pixelWidth = Math.max(Math.round(cssWidth * dpr), 1);
    const pixelHeight = Math.max(Math.round(cssHeight * dpr), 1);
    if (this.cache.width === pixelWidth && this.cache.height === pixelHeight) return;

    this.cache.width = pixelWidth;
    this.cache.height = pixelHeight;
    this.cacheWidth = cssWidth;
    this.cacheHeight = cssHeight;
    // Resizing a canvas resets its context, so the DPR transform is reapplied
    // here. Everything downstream can then work in CSS pixels.
    this.cacheCtx.setTransform(dpr, 0, 0, dpr, 0, 0);
    this.cacheStale = true;
    this.version += 1;
  }

  // ---- stroke lifecycle -------------------------------------------------

  beginStroke(handId: number, tool: Tool, color: string, width: number, at: BoardPoint): void {
    this.active.set(handId, { kind: 'stroke', tool, color, width, points: [at] });
    this.version += 1;
  }

  /** @returns false when the sample was rejected, which ends the stroke. */
  extendStroke(handId: number, at: BoardPoint): boolean {
    const stroke = this.active.get(handId);
    if (!stroke) return false;

    const last = stroke.points[stroke.points.length - 1];
    const distance = Math.hypot(at.x - last.x, at.y - last.y);

    if (distance > STROKE.maxJumpDistance) return false;
    if (distance < STROKE.minPointDistance) return true;

    stroke.points.push(at);
    this.version += 1;
    return true;
  }

  endStroke(handId: number): void {
    const stroke = this.active.get(handId);
    if (!stroke) return;
    this.active.delete(handId);
    this.pushAction(stroke);
  }

  cancelStroke(handId: number): void {
    if (this.active.delete(handId)) this.version += 1;
  }

  endAllStrokes(): void {
    for (const handId of [...this.active.keys()]) this.endStroke(handId);
  }

  // ---- history ----------------------------------------------------------

  clear(): void {
    this.endAllStrokes();
    // Nothing on screen means nothing to clear — do not stack redundant clears.
    if (this.stats.strokes === 0) return;
    // Recorded as an action so that clearing is undoable like anything else.
    this.pushAction({ kind: 'clear' });
  }

  undo(): void {
    this.endAllStrokes();
    const action = this.actions.pop();
    if (!action) return;
    this.redoStack.push(action);
    this.cacheStale = true;
    this.version += 1;
    this.notify();
  }

  redo(): void {
    const action = this.redoStack.pop();
    if (!action) return;
    this.actions.push(action);
    this.applyToCache(action);
    this.version += 1;
    this.notify();
  }

  private pushAction(action: BoardAction): void {
    this.actions.push(action);
    this.redoStack.length = 0;
    this.applyToCache(action);
    this.version += 1;
    this.notify();
  }

  /** Appending never needs a replay — only removal does. */
  private applyToCache(action: BoardAction): void {
    if (this.cacheStale) return;
    if (!this.cacheMapper) {
      this.cacheStale = true;
      return;
    }
    if (action.kind === 'clear') {
      this.cacheCtx.clearRect(0, 0, this.cacheWidth, this.cacheHeight);
    } else {
      paintStroke(this.cacheCtx, action, this.cacheMapper);
    }
  }

  private notify(): void {
    this.onChange?.(this.stats);
  }

  // ---- rendering --------------------------------------------------------

  /**
   * Draws finished ink plus any in-progress strokes onto the visible ink layer.
   * All dimensions are CSS pixels; both contexts carry the DPR transform.
   */
  render(ctx: CanvasRenderingContext2D, mapper: FrameMapper, width: number, height: number): void {
    this.cacheMapper = mapper;

    if (this.cacheStale) {
      paintActions(this.cacheCtx, this.actions, mapper, this.cacheWidth, this.cacheHeight);
      this.cacheStale = false;
    }

    ctx.clearRect(0, 0, width, height);
    if (this.cache.width > 0 && this.cache.height > 0) {
      ctx.drawImage(this.cache, 0, 0, this.cacheWidth, this.cacheHeight);
    }
    for (const stroke of this.active.values()) paintStroke(ctx, stroke, mapper);
  }

  /** Forces a replay on the next render — call after a resize or a mapper change. */
  invalidate(): void {
    this.cacheStale = true;
    this.version += 1;
  }
}

File: drawone/board/board.py
"""The drawing: a list of vector actions plus a rasterised cache of the finished ones."""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Callable, Dict, Iterable, List, Optional, Sequence

from ..config import STROKE
from ..render.canvas import Layer, Point, flatten_quadratic, stroke_polyline
from ..render.color import to_bgr
from ..tracking.frame_mapper import FrameMapper
from ..types import BoardAction, BoardPoint, ClearAction, StrokeAction, Tool


def build_path(stroke: StrokeAction, mapper: FrameMapper) -> List[Point]:
    """Traces a stroke as a chain of quadratic curves whose control points are
    the samples and whose endpoints are the midpoints between them.

    Two things fall out of that: the curve is C1 continuous, so the line reads as
    one gesture rather than a polyline, and widely spaced samples during a fast
    movement are bridged by a curve instead of leaving a gap.
    """
    pts = [mapper.board_to_stage(p.x, p.y) for p in stroke.points]
    n = len(pts)

    if n == 1:
        return [(pts[0].x, pts[0].y)]
    if n == 2:
        return [(pts[0].x, pts[0].y), (pts[1].x, pts[1].y)]

    path: List[Point] = [(pts[0].x, pts[0].y)]
    current = pts[1]
    for i in range(1, n - 1):
        following = pts[i + 1]
        midpoint = (
            (current.x + following.x) / 2,
            (current.y + following.y) / 2,
        )
        flatten_quadratic(path[-1], (current.x, current.y), midpoint, path)
        current = following
    path.append((current.x, current.y))
    return path


def paint_stroke(layer: Layer, stroke: StrokeAction, mapper: FrameMapper) -> None:
    width = max(stroke.width * mapper.pixels_per_unit, 0.6)
    stroke_polyline(
        layer,
        build_path(stroke, mapper),
        width,
        to_bgr(stroke.color),
        erase=stroke.tool == "eraser",
    )


def paint_actions(
    layer: Layer, actions: Sequence[BoardAction], mapper: FrameMapper
) -> None:
    layer.clear()
    for action in actions:
        if isinstance(action, ClearAction):
            layer.clear()
        else:
            paint_stroke(layer, action, mapper)


@dataclass(frozen=True)
class BoardStats:
    strokes: int
    can_undo: bool
    can_redo: bool


class Board:
    """Holds the drawing as a list of vector actions and keeps a rasterised cache
    of the finished ones.

    Finished strokes are painted into the cache exactly once; only in-progress
    strokes are re-traced each frame. Undo, redo, clear and window resizes replay
    the action list, which is also what makes undo work across an erase or a
    clear.
    """

    def __init__(self) -> None:
        self._actions: List[BoardAction] = []
        self._redo: List[BoardAction] = []
        self._active: Dict[int, StrokeAction] = {}

        self._cache = Layer(1, 1)
        self._scratch = Layer(1, 1)
        self._cache_stale = True
        #: The mapper the cache was rasterised with; None until the first render.
        self._cache_mapper: Optional[FrameMapper] = None

        #: Bumped whenever the visible result would differ.
        self.version = 0
        self.on_change: Optional[Callable[[BoardStats], None]] = None

    @property
    def stats(self) -> BoardStats:
        """`strokes` counts what is currently *visible*, not everything ever
        drawn — a clear resets it to zero. Clear and Save key off this, and both
        should be dead on an empty board even though the history behind it is
        still intact.
        """
        visible = 0
        for action in self._actions:
            if isinstance(action, ClearAction):
                visible = 0
            else:
                visible += 1
        return BoardStats(
            strokes=visible,
            can_undo=len(self._actions) > 0,
            can_redo=len(self._redo) > 0,
        )

    @property
    def active_strokes(self) -> Iterable[StrokeAction]:
        return list(self._active.values())

    @property
    def has_active_strokes(self) -> bool:
        return len(self._active) > 0

    def has_active_stroke(self, hand_id: int) -> bool:
        return hand_id in self._active

    def get_actions(self) -> Sequence[BoardAction]:
        return self._actions

    def resize(self, width: int, height: int) -> None:
        if not self._cache.resize(width, height):
            return
        self._scratch.resize(width, height)
        self._cache_stale = True
        self.version += 1

    # ---- stroke lifecycle -------------------------------------------------

    def begin_stroke(
        self, hand_id: int, tool: Tool, color: str, width: float, at: BoardPoint
    ) -> None:
        self._active[hand_id] = StrokeAction(
            tool=tool, color=color, width=width, points=[at]
        )
        self.version += 1

    def extend_stroke(self, hand_id: int, at: BoardPoint) -> bool:
        """Returns False when the sample was rejected, which ends the stroke."""
        stroke = self._active.get(hand_id)
        if stroke is None:
            return False

        last = stroke.points[-1]
        distance = math.hypot(at.x - last.x, at.y - last.y)

        if distance > STROKE.max_jump_distance:
            return False
        if distance < STROKE.min_point_distance:
            return True

        stroke.points.append(at)
        self.version += 1
        return True

    def end_stroke(self, hand_id: int) -> None:
        stroke = self._active.pop(hand_id, None)
        if stroke is None:
            return
        self._push_action(stroke)

    def cancel_stroke(self, hand_id: int) -> None:
        if self._active.pop(hand_id, None) is not None:
            self.version += 1

    def end_all_strokes(self) -> None:
        for hand_id in list(self._active.keys()):
            self.end_stroke(hand_id)

    # ---- history ----------------------------------------------------------

    def clear(self) -> None:
        self.end_all_strokes()
        # Nothing on screen means nothing to clear — do not stack redundant clears.
        if self.stats.strokes == 0:
            return
        # Recorded as an action so that clearing is undoable like anything else.
        self._push_action(ClearAction())

    def undo(self) -> None:
        self.end_all_strokes()
        if not self._actions:
            return
        self._redo.append(self._actions.pop())
        self._cache_stale = True
        self.version += 1
        self._notify()

    def redo(self) -> None:
        if not self._redo:
            return
        action = self._redo.pop()
        self._actions.append(action)
        self._apply_to_cache(action)
        self.version += 1
        self._notify()

    def _push_action(self, action: BoardAction) -> None:
        self._actions.append(action)
        self._redo.clear()
        self._apply_to_cache(action)
        self.version += 1
        self._notify()

    def _apply_to_cache(self, action: BoardAction) -> None:
        """Appending never needs a replay — only removal does."""
        if self._cache_stale:
            return
        if self._cache_mapper is None:
            self._cache_stale = True
            return
        if isinstance(action, ClearAction):
            self._cache.clear()
        else:
            paint_stroke(self._cache, action, self._cache_mapper)

    def _notify(self) -> None:
        if self.on_change is not None:
            self.on_change(self.stats)

    # ---- rendering --------------------------------------------------------

    def render(self, mapper: FrameMapper, width: int, height: int) -> Layer:
        """Returns the layer to composite: finished ink plus any live strokes.

        With nothing in progress this is the cache itself, so a settled drawing
        costs no per-frame raster work at all.
        """
        self.resize(width, height)
        self._cache_mapper = mapper

        if self._cache_stale:
            paint_actions(self._cache, self._actions, mapper)
            self._cache_stale = False

        if not self._active:
            return self._cache

        self._scratch.copy_from(self._cache)
        for stroke in self._active.values():
            paint_stroke(self._scratch, stroke, mapper)
        return self._scratch

    def invalidate(self) -> None:
        """Forces a replay on the next render — after a resize or a mapper change."""
        self._cache_stale = True
        self.version += 1
